### PR TITLE
Fix Nirspec blotting to support outlier detection:  JP-1569

### LINKED ISSUE #5130: Implement NIRSpec IFU blotting algorithm
_Issue [JP-1569](https://jira.stsci.edu/browse/JP-1569) was created on JIRA by [Jonathan Eisenhamer](https://jira.stsci.edu/secure/ViewProfile.jspa?name=eisenham):_

Implement new NIRSPec IFU blotting algorithm. This ticket follows from JP-1405.
Outlier detection calls cube_build to 'blot' a median sky image back to the detector. In 
blot_cube_build the invert wcs function of taking the ra, dec, and wavelength values and
inverting those values to the detector does not work correctly. For NIRSPEC this operation has to be performed slice by slice. The error is that each position on the sky in mapped back to several slices (> 10) which is not possible. 
This ticket aims at storing the sky footprint of each slice when cube_build is initially called by outlier detection (when 'single' cube are built). This footprint is passed back to outlier detection which them passes the information back to blot_cube_build.

### PR DESCRIPTION
**Pull request by jemorrison**
[Could use some supporting descriptive info here]

Fixes #5130 / [JP-1569](https://jira.stsci.edu/browse/JP-1569)

### CODE REVIEW

File: CHANGES.rst [hunk 1]
```diff
@@ -33,6 +33,7 @@ combine_1d
 
 cube_build
 ----------
+- modified NIRSpec blotting find min and max ra and dec for each slice and only invert those values on slice that fall in range [#5144]
```
Comment (hbushouse) [inline CHANGES.rst:36]: Is there a verb missing here somewhere, such as "... blotting **TO** find min and max ra and dec ..."?

File: jwst/cube_build/blot_cube_build.py [hunk 2]
```diff
@@ -281,19 +231,135 @@ def blot_overlap_quick(self, model, xcenter, ycenter, xstart,
...
+                dxy = [ (dx*dx + dy*dy)  for dy in d2pix for dx in d1pix]
                 dxy = np.sqrt(dxy)
                 weight_distance = np.exp(-dxy)
                 weighted_flux = weight_distance * flux_cube[ipt]
                 index2d = [iy * blot_xsize + ix for iy in index_y[0] for ix in (index_x[0]+xstart)]
                 blot_flux[index2d] = blot_flux[index2d] + weighted_flux
                 blot_weight[index2d] = blot_weight[index2d] + weight_distance
 
+        t1 = time.time()
+        log.debug("Time to blot median image to input model = %.1f s" % (t1-t0,))
```
Comment (hbushouse) [inline jwst/cube_build/blot_cube_build.py:244]: Not a big deal, but could change this to use the new python string formatting syntax:
```
log.debug(f"Time to blot median image to input model = {t1-t0:.1f}")
```

File: jwst/cube_build/blot_cube_build.py [hunk 3]
```diff
@@ -281,19 +231,135 @@ def blot_overlap_quick(self, model, xcenter, ycenter, xstart,
...
         igood = np.where(blot_weight > 0)
         blot_flux[igood] = blot_flux[igood] / blot_weight[igood]
         blot_flux = blot_flux.reshape((blot_ysize, blot_xsize))
         return blot_flux
+
+    # ************************************************************************
+    def blot_images_nirspec(self):
+        """ Core blotting routine for NIRSPEC
+
+        This is the main routine for blotting the NIRSPEC median sky image back to
```
Comment (hbushouse) [inline jwst/cube_build/blot_cube_build.py:255]: Technically isn't it a median sky **cube** (as opposed to image)?

File: jwst/cube_build/blot_cube_build.py [hunk 4]
```diff
@@ -281,19 +231,135 @@ def blot_overlap_quick(self, model, xcenter, ycenter, xstart,
...
+        the detector space and creating a blotting image for each input model.
+        This routine was split from the MIRI routine because the blotting for NIRSpec
+        needs to be done slice by slice and an error in the inverse mapping (sky to
+        detector) mapped too many values back to the detector. This routine adds
+        a check and first pulls out the min and max ra and dec values in the slice
+        and only inverts the slice values back to the detector.
+        1. Loop over every data model to be blotted and find ra,dec,wavelength
+           for every pixel in a valid slice on the detector.
+        2. For each slice get the WCS info for the slice. Find the x,y bounding
```
Comment (hbushouse) [inline jwst/cube_build/blot_cube_build.py:264]: It sounds like ra, dec, wavelength values were found in step 1. Are they being found/computed again in step 2? Or just using the values from step 1 to find the min/max?

File: jwst/cube_build/blot_cube_build.py [hunk 5]
```diff
@@ -281,19 +231,135 @@ def blot_overlap_quick(self, model, xcenter, ycenter, xstart,
...
+            blot.dq = None
+            # ___________________________________________________________________
+            ysize, xsize = model.data.shape
+            ycenter = np.arange(ysize)
+            xcenter = np.arange(xsize)
+
+            # for NIRSPEC wcs information accessed seperately for each slice
+            nslices = 30
+            log.info('Looping over 30 slices on NIRSPEC detector, this takes a little while')
+            t0 = time.time()
```
Comment (hbushouse) [inline jwst/cube_build/blot_cube_build.py:293]: t0 got set up in line 281 and now it's being set again.

File: jwst/cube_build/blot_cube_build.py [hunk 6]
```diff
@@ -281,19 +231,135 @@ def blot_overlap_quick(self, model, xcenter, ycenter, xstart,
...
+            roi_det = 1.0  # Just large enough that we don't get holes
+            for ii in range(nslices):
+                ts0 = time.time()
+                # for each slice pull out the blotted values that actually fall on the slice region
+                # used the bounding box of each slice to determine the slice limits
```
Comment (hbushouse) [inline jwst/cube_build/blot_cube_build.py:298]: "use" instead of "used"?

File: jwst/cube_build/blot_cube_build.py [hunk 7]
```diff
@@ -281,19 +231,135 @@ def blot_overlap_quick(self, model, xcenter, ycenter, xstart,
...
+                yuse = np.logical_and(y_slice >= ylimit[0], y_slice <= ylimit[1])
+
+                fuse = np.logical_and(xuse, yuse)
+                x_slice = x_slice[fuse]
+                y_slice = y_slice[fuse]
+                flux_slice = flux_slice[fuse]
+
+                nn = flux_slice.size
+                for ipt in range(nn):
+                    # search xcenter and ycenter seperately. These arrays are smallsh.
```
Comment (hbushouse) [inline jwst/cube_build/blot_cube_build.py:335]: typo: "small**i**sh"

File: jwst/cube_build/blot_cube_build.py [hunk 8]
```diff
@@ -281,19 +231,135 @@ def blot_overlap_quick(self, model, xcenter, ycenter, xstart,
...
+
+                        dxy = [ (dx*dx + dy*dy)  for dy in d2pix for dx in d1pix]
+                        dxy = np.sqrt(dxy)
+                        weight_distance = np.exp(-dxy)
+                        weighted_flux = weight_distance * flux_slice[ipt]
+                        index2d = [iy * blot_xsize + ix for iy in index_y[0] for ix in (index_x[0])]
+                        blot_flux[index2d] = blot_flux[index2d] + weighted_flux
+                        blot_weight[index2d] = blot_weight[index2d] + weight_distance
+                ts1 = time.time()
+                log.debug("Time to map 1 slice  = %.1f s" % (ts1-ts0,))
```
Comment (hbushouse) [inline jwst/cube_build/blot_cube_build.py:356]:
```
log.debug(f"Time to map 1 slice = {ts1-ts0:.1f}")
```

File: jwst/cube_build/blot_cube_build.py [hunk 9]
```diff
@@ -281,19 +231,135 @@ def blot_overlap_quick(self, model, xcenter, ycenter, xstart,
...
+            # done mapping median cube  to this input model
+            t1 = time.time()
+            log.debug("Time to blot median image to input model = %.1f s" % (t1-t0,))
```
Comment (hbushouse) [inline jwst/cube_build/blot_cube_build.py:359]:
```
log.debug(f"Time to blot median image to input model = {t1-t0:.1f}")
```

File: jwst/cube_build/cube_build_step.py [hunk 10]
```diff
@@ -292,6 +292,10 @@ def process(self, input):
             self.log.info('Number of ifucubes produced by a this run %i',
                           num_cubes)
 
+        if self.single:
+            self.log.info('Number of single ifucubes produced by a this run %i',
```
Comment (hbushouse) [inline jwst/cube_build/cube_build_step.py:296]: typo: "... produced by a this ..."

Comment (hbushouse) [reply jwst/cube_build/cube_build_step.py:296]: Or just replace with:
```
self.log.info(f'Number of single ifucubes produced by this run = {num_cubes}')
```

File: jwst/cube_build/ifu_cube.py [hunk 11]
```diff
@@ -20,6 +20,7 @@
 from . import coord
 
 log = logging.getLogger(__name__)
+log.addHandler(logging.StreamHandler())
```
Comment (hbushouse) [inline jwst/cube_build/ifu_cube.py:23]: Just curious as to what ``StreamHandler`` is used for?

Comment (jemorrison) [reply jwst/cube_build/ifu_cube.py:23]: I had to add StreamHandler to the code to get my debug statements to print the screen. Jonathan suggested this when I did not get any of my debug statements to print out. I do hope I understood him correctly. I now get the debug statements to print out, but they always print out even when I do not have debug set up by using  --debug.
So how do you (@hbushouse) control printing debug statements in code ?

File: jwst/cube_build/blot_cube_build.py (hunk 3 above)
Comment (jemorrison) [reply jwst/cube_build/blot_cube_build.py]: fixed

File: jwst/cube_build/blot_cube_build.py (hunk 2 above)
Comment (jemorrison) [reply jwst/cube_build/blot_cube_build.py]: fixed

File: CHANGES.rst (hunk 1 above)
Comment (jemorrison) [reply CHANGES.rst]: fixed

File: jwst/cube_build/blot_cube_build.py (hunk 4 above)
Comment (jemorrison) [reply jwst/cube_build/blot_cube_build.py]: rewrote section

File: jwst/cube_build/blot_cube_build.py (hunk 5 above)
Comment (jemorrison) [reply jwst/cube_build/blot_cube_build.py:292]: removed one of the t0

File: jwst/cube_build/blot_cube_build.py (hunk 7 above)
Comment (jemorrison) [reply jwst/cube_build/blot_cube_build.py]: fixed

File: jwst/cube_build/cube_build_step.py (hunk 10 above)
Comment (jemorrison) [reply jwst/cube_build/cube_build_step.py]: fixed

**Comment by jemorrison** [comment]
@hbushouse I believe it is ready to be merged

File: jwst/cube_build/ifu_cube.py (hunk 11 above)
Comment (hbushouse) [reply jwst/cube_build/ifu_cube.py:23]: Huh, I've never encountered that problem. Usually just having line 24 (`log.setLevel(logging.DEBUG)`) in conjunction with specifying "-v" on the command line results in all the debug messages appearing. Or are you talking about from within a python session (as opposed to command line)?

Comment (jemorrison) [reply jwst/cube_build/ifu_cube.py:23]: I took out the StreamHandler and now the debug works for me. I don't know why it was not before.

**Pull request merged**